plugin/src/main/java/org/opensearch/ml/processor/MLInferenceIngestProcessor.java

```diff
@@ -0,0 +1,368 @@
+import static org.opensearch.ml.processor.InferenceProcessorAttributes.MODEL_CONFIG;
+import static org.opensearch.ml.processor.InferenceProcessorAttributes.MODEL_ID;
+import static org.opensearch.ml.processor.InferenceProcessorAttributes.OUTPUT_MAP;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.BiConsumer;
+
+import org.opensearch.action.ActionRequest;
+import org.opensearch.client.Client;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.core.common.Strings;
+import org.opensearch.ingest.AbstractProcessor;
+import org.opensearch.ingest.ConfigurationUtils;
+import org.opensearch.ingest.IngestDocument;
+import org.opensearch.ingest.Processor;
+import org.opensearch.ingest.ValueSource;
+import org.opensearch.ml.common.output.model.ModelTensorOutput;
+import org.opensearch.ml.common.transport.MLTaskResponse;
+import org.opensearch.ml.common.transport.prediction.MLPredictionTaskAction;
+import org.opensearch.script.ScriptService;
+import org.opensearch.script.TemplateScript;
+
+import com.jayway.jsonpath.Configuration;
+import com.jayway.jsonpath.JsonPath;
+import com.jayway.jsonpath.Option;
+
+/**
+ * MLInferenceIngestProcessor requires a modelId string to call model inferences
+ * maps fields in document for model input, and maps model inference output to new document fields
+ * this processor also handles dot path notation for nested object( map of array) by rewriting json path accordingly
+ */
+public class MLInferenceIngestProcessor extends AbstractProcessor implements ModelExecutor {
+
+    private final InferenceProcessorAttributes inferenceProcessorAttributes;
+    private final boolean ignoreMissing;
+    private final ScriptService scriptService;
+    private static Client client;
+    public static final String TYPE = "ml_inference";
+    public static final String DEFAULT_OUTPUT_FIELD_NAME = "inference_results";
+
+    protected MLInferenceIngestProcessor(
+        String modelId,
+        List<Map<String, String>> inputMaps,
+        List<Map<String, String>> outputMaps,
+        Map<String, String> modelConfig,
+        String tag,
+        String description,
+        boolean ignoreMissing,
+        ScriptService scriptService,
+        Client client
+    ) {
+        super(tag, description);
+        this.inferenceProcessorAttributes = new InferenceProcessorAttributes(modelId, inputMaps, outputMaps, modelConfig);
+        this.ignoreMissing = ignoreMissing;
+        this.scriptService = scriptService;
+        this.client = client;
+    }
+
+    /**
+     * overwrite in this execute method,
+     * when batch inference is available,
+     * to support async multiple predictions.
+     */
+    @Override
+    public void execute(IngestDocument ingestDocument, BiConsumer<IngestDocument, Exception> handler) {
```

Comment:
Unlike the synchronous execute() that returns a IngestDocument, this async method does not return anything. So it means this processor doesn't return anything after the execution but only relies on the async handler to process. Does it mean this processor cannot be concatenated with other handlers, or in other words, it can only serve as the last processor in an ingest pipeline?

Reply:
this processor extends from the [Processor](https://github.com/opensearch-project/OpenSearch/blob/db694a9a486ba6d8f39bf11136da0c9220edaf0d/server/src/main/java/org/opensearch/ingest/Processor.java#L65) interface, in here you can tell, the void execute() will run IngestDocument  execute() inside. And you are right that it's only only on the async handler to process.

Reply:
> And you are right that it's only only on the async handler to process.

@mingshl ,  you mean this part is right `Does it mean this processor cannot be concatenated with other handlers, or in other words, it can only serve as the last processor in an ingest pipeline?` ?

Reply:
it can be work chaining with other processors. @Zhangxunmt  is mostly concerning whether the ingest processors can be async